contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
 	PG_RETURN_NUMERIC(result);
 }
 
+Datum
+parsename(PG_FUNCTION_ARGS)
+{
+	text *object_name = PG_GETARG_TEXT_PP(0);
+	int object_piece = PG_GETARG_INT32(1);
+	char *object_name_str = text_to_cstring(object_name);
+	UErrorCode errorCode = U_ZERO_ERROR;
+	UChar unicode_object_name_str[256] = {0};
+	int len;
+	UChar requested_object_part[256] = {0};
+	int current_part = 1;
+	int state = 0; // 0: outside quotes and brackets, 1: inside quotes, 2: inside brackets
+	int improperly_enclosed = 0;
+	int max_parts = 1;
+	int max_length;
+	int part_lengths[256] = {0};
+	
+	// Calculate max_parts based on the number of dots in the input string
+	for (char *dot = strchr(object_name_str, '.'); dot != NULL; dot = strchr(dot + 1, '.'))
```

Comment:
This is wrong, dot is not always delimiter of parts.
It could be part of the name as well. This error should be detected by your test if you add all tests for special characters
1> select parsename('[abc.abc]', 1);
2> go

--------------------------------------------------------------------------------------------------------------------------------
abc.abc